File: frontend/src/types/interviewQuestion.ts
// PrepBench - Copyright (c) 2026 Nimish Kanungo
// Licensed under the PolyForm Noncommercial License 1.0.0 (see LICENSE).
// Commercial use requires a separate licence from the copyright holder.

export type InterviewRoundType = 'hr_screening' | 'hiring_manager' | 'system_design' | 'behavioral';

export interface RoundTypeInfo {
  value: InterviewRoundType;
  label: string;
  /** How long a good answer to this round runs, in seconds. */
  target_min_seconds?: number;
  target_max_seconds?: number;
  /** Suggested thinking time before answering. */
  thinking_seconds?: number;
  /** A shape to plan the answer with. */
  plan_prompt?: string;
  /** What the interviewer is listening for. Guidance, not a grade. */
  listening_for?: string;
  /** The rubric an analysis grades content against. */
  content_categories?: string[];
}

export interface InterviewQuestion {
  id: number;
  round_type: InterviewRoundType;
  question_text: string;
  category: string | null;
  prepared_answer?: string | null;
  key_talking_points?: string[] | null;
  is_ai_generated: boolean;
  created_at: string;
  /** Takes recorded of this question. Filled by the list endpoint. */
  practice_count?: number;
}

export interface GenerateInterviewQuestionRequest {
  round_type: InterviewRoundType;
  topic?: string;
  save_to_bank?: boolean;
}

export interface InterviewQuestionUpdate {
  question_text?: string;
  category?: string;
  prepared_answer?: string | null;
  key_talking_points?: string[] | null;
}

export interface InterviewQuestionImportResult {
  imported_count: number;
  skipped_count: number;
  errors: string[];
}

export interface ImportInterviewQuestionsRequest {
  defaultRoundType: InterviewRoundType;
  defaultCategory?: string;
  file?: File;
  text?: string;
}

File: backend/app/schemas/interview_question.py
# PrepBench - Copyright (c) 2026 Nimish Kanungo
# Licensed under the PolyForm Noncommercial License 1.0.0 (see LICENSE).
# Commercial use requires a separate licence from the copyright holder.

from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field
from app.models.interview_question import InterviewRoundType


class InterviewQuestionBase(BaseModel):
    round_type: InterviewRoundType
    question_text: str
    category: Optional[str] = None
    prepared_answer: Optional[str] = None
    key_talking_points: Optional[List[str]] = None


class InterviewQuestionCreate(InterviewQuestionBase):
    is_ai_generated: bool = False
    source_topic: Optional[str] = None


class InterviewQuestionResponse(InterviewQuestionBase):
    id: int
    is_ai_generated: bool
    created_at: datetime
    # How many takes of this question have been recorded. Filled by the list
    # endpoint; 0 elsewhere rather than a guess.
    practice_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class InterviewQuestionUpdate(BaseModel):
    question_text: Optional[str] = None
    category: Optional[str] = None
    prepared_answer: Optional[str] = None
    key_talking_points: Optional[List[str]] = None


class InterviewQuestionImportResult(BaseModel):
    imported_count: int
    skipped_count: int
    errors: List[str] = []


class InterviewQuestionFilter(BaseModel):
    round_type: Optional[InterviewRoundType] = None
    category: Optional[str] = None
    keyword: Optional[str] = None


class GenerateInterviewQuestionRequest(BaseModel):
    round_type: InterviewRoundType
    topic: Optional[str] = None
    save_to_bank: bool = False


class RoundTypeInfo(BaseModel):
    value: str
    label: str
    # Guidance for answering this round -- see services/interview_rounds.py.
    # Defaults keep the shape valid for any caller that builds one bare.
    target_min_seconds: int = 0
    target_max_seconds: int = 0
    thinking_seconds: int = 0
    plan_prompt: str = ""
    listening_for: str = ""
    content_categories: List[str] = []
